src/LinearMap.jl

```diff
@@ -21,106 +21,165 @@ independent of the numeric type of the wrapped set (`N`).
+
+julia> lm2 = A * X
+LinearMap{Int64,BallInf{Int64},Int64,Array{Int64,2}}([1 2; 1 3; 1 4], BallInf{Int64}([0, 0], 1))
+
+julia> lm == lm2
+true
+```
+
+For convenience, `A` does not need to be a matrix but we also allow to use
+vectors (interpreted as an ``n×1`` matrix) and `UniformScaling`s resp. scalars
+(interpreted as a scaling, i.e., a scaled identity matrix).
+Scaling by ``1`` is ignored.
+
+```jldoctest constructors
+julia> using LinearAlgebra: I
+
+julia> [2, 3] * Y
+LinearMap{Int64,BallInf{Int64},Int64,Array{Int64,2}}([2; 3], BallInf{Int64}([0], 1))
+
+julia> lm3 = 2 * X
+LinearMap{Int64,BallInf{Int64},Int64,SparseArrays.SparseMatrixCSC{Int64,Int64}}(
+  [1, 1]  =  2
+  [2, 2]  =  2, BallInf{Int64}([0, 0], 1))
+
+julia> 2I * X == lm3
+true
+
+julia> 1I * X == X
+true
+```
+
+Applying a linear map to a `LinearMap` object combines the two maps into a
+single `LinearMap` instance.
+Again we can make use of the conversion for convenience.
+
+```jldoctest constructors
+julia> B = transpose(A); B * lm
+LinearMap{Int64,BallInf{Int64},Int64,Array{Int64,2}}([3 9; 9 29], BallInf{Int64}([0, 0], 1))
+
+julia> B = [3, 4, 5]; B * lm
+LinearMap{Int64,BallInf{Int64},Int64,Array{Int64,2}}([12 38], BallInf{Int64}([0, 0], 1))
+
+julia> B = 2; B * lm
+LinearMap{Int64,BallInf{Int64},Int64,Array{Int64,2}}([2 4; 2 6; 2 8], BallInf{Int64}([0, 0], 1))
+```
+
+The application of a `LinearMap` to a `ZeroSet` or an `EmptySet` is simplified
+automatically.
+
+```jldoctest constructors
+julia> A * ZeroSet{Int}(2)
+ZeroSet{Int64}(3)
+
+julia> A * EmptySet{Int}()
+EmptySet{Int64}()
+```
 """
 struct LinearMap{N<:Real, S<:LazySet{N},
                  NM, MAT<:AbstractMatrix{NM}} <: LazySet{N}
     M::MAT
     X::S
 
     # default constructor with dimension match check
-    function LinearMap{N, S, NM, MAT}(M::MAT, X::S) where {N<:Real,
-            S<:LazySet{N}, NM, MAT<:AbstractMatrix{NM}}
+    function LinearMap(M::MAT, X::S) where {N<:Real, S<:LazySet{N}, NM,
+                                            MAT<:AbstractMatrix{NM}}
         @assert dim(X) == size(M, 2) "a linear map of size $(size(M)) cannot " *
             "be applied to a set of dimension $(dim(X))"
         return new{N, S, NM, MAT}(M, X)
     end
 end
 
-# convenience constructor without type parameter
-LinearMap(M::MAT,
-          X::S) where {N<:Real, S<:LazySet{N}, NM, MAT<:AbstractMatrix{NM}} =
-    LinearMap{N, S, NM, MAT}(M, X)
-
-# constructor from a linear map: perform the matrix multiplication immediately
-LinearMap(M::MAT, lm::LinearMap{N, S, NM, MAT}) where {N<:Real, S<:LazySet{N},
-        NM, MAT<:AbstractMatrix{NM}} =
-    LinearMap{N, S, NM, MAT}(M * lm.M, lm.X)
-
 """
 ```
-    *(M::AbstractMatrix{N}, X::LazySet{N}) where {N<:Real}
+    *(map, X::LazySet)
 ```
 
-Return the linear map of a convex set.
+Alias to create a `LinearMap` object.
 
 ### Input
 
-- `M` -- matrix/linear map
-- `X` -- convex set
+- `map` -- linear map
```

Comment:
I don't understand the change from `M::AbstractMatrix`  to "map". This function is meant to produce a new lazy linear map form a given matrix and set type. Other uses should be added through dispatch as usual (or eventually with a small `Union` type).

Reply:
I recently saw a Julia talk about dispatch. One of the messages was that not adding types is helpful for new implementations to "just work". This was my motivation. I can remove these methods again if you prefer.

Reply:
I had a closer look: the correct way to do things here is to add dispatch methods of `LinearMap`. We currently have them for `*`. Now we only need a single method for `*`, so the old ones become redundant. While they do not hurt, but I prefer to have a smaller code base. So I had to combine all of them into a single method. I did not want to have a big `Union` argument, but changed it now as per your request.